File: src/cstlib/dynamics.py
"""Deterministic nonlinear systems for CST experiments."""
from dataclasses import dataclass

@dataclass
class Lorenz:
    sigma: float = 10.0
    rho: float = 28.0
    beta: float = 8.0 / 3.0
    x: float = 1.0
    y: float = 1.0
    z: float = 1.0

    def _f(self, x: float, y: float, z: float) -> tuple[float, float, float]:
        return self.sigma * (y - x), x * (self.rho - z) - y, x * y - self.beta * z

    def step(self, dt: float = 0.01) -> tuple[float, float, float]:
        if dt <= 0:
            raise ValueError("dt must be positive")
        x, y, z = self.x, self.y, self.z
        k1 = self._f(x, y, z)
        k2 = self._f(x + dt*k1[0]/2, y + dt*k1[1]/2, z + dt*k1[2]/2)
        k3 = self._f(x + dt*k2[0]/2, y + dt*k2[1]/2, z + dt*k2[2]/2)
        k4 = self._f(x + dt*k3[0], y + dt*k3[1], z + dt*k3[2])
        self.x += dt*(k1[0]+2*k2[0]+2*k3[0]+k4[0])/6
        self.y += dt*(k1[1]+2*k2[1]+2*k3[1]+k4[1])/6
        self.z += dt*(k1[2]+2*k2[2]+2*k3[2]+k4[2])/6
        return self.x, self.y, self.z

    def snapshot(self) -> dict[str, float]:
        return {"x": self.x, "y": self.y, "z": self.z}

File: cosmic rust/src/dynamics.rs
#[derive(Clone, Debug, PartialEq)]
pub struct Lorenz {
    pub sigma: f64,
    pub rho: f64,
    pub beta: f64,
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

impl Default for Lorenz {
    fn default() -> Self {
        Self {
            sigma: 10.0,
            rho: 28.0,
            beta: 8.0 / 3.0,
            x: 1.0,
            y: 1.0,
            z: 1.0,
        }
    }
}

impl Lorenz {
    fn derivative(&self, x: f64, y: f64, z: f64) -> (f64, f64, f64) {
        (
            self.sigma * (y - x),
            x * (self.rho - z) - y,
            x * y - self.beta * z,
        )
    }

    pub fn step(&mut self, dt: f64) -> Result<(f64, f64, f64), String> {
        if dt <= 0.0 {
            return Err("dt must be positive".into());
        }
        let (x, y, z) = (self.x, self.y, self.z);
        let k1 = self.derivative(x, y, z);
        let k2 = self.derivative(
            x + dt * k1.0 / 2.0,
            y + dt * k1.1 / 2.0,
            z + dt * k1.2 / 2.0,
        );
        let k3 = self.derivative(
            x + dt * k2.0 / 2.0,
            y + dt * k2.1 / 2.0,
            z + dt * k2.2 / 2.0,
        );
        let k4 = self.derivative(x + dt * k3.0, y + dt * k3.1, z + dt * k3.2);
        self.x += dt * (k1.0 + 2.0 * k2.0 + 2.0 * k3.0 + k4.0) / 6.0;
        self.y += dt * (k1.1 + 2.0 * k2.1 + 2.0 * k3.1 + k4.1) / 6.0;
        self.z += dt * (k1.2 + 2.0 * k2.2 + 2.0 * k3.2 + k4.2) / 6.0;
        Ok((self.x, self.y, self.z))
    }
}
